src/Components/Server/src/ComponentHub.cs

```diff
@@ -156,104 +172,105 @@ public string StartCircuit(string baseUri, string uri)
             {
-                Log.CircuitHostNotInitialized(_logger);
-                _ = NotifyClientError(Clients.Caller, "Circuit not initialized.");
                 return;
             }
 
             _ = CircuitHost.EndInvokeJSFromDotNet(asyncHandle, succeeded, arguments);
         }
 
-        /// <summary>
-        /// Intended for framework use only. Applications should not call this method directly.
-        /// </summary>
-        public void DispatchBrowserEvent(string eventDescriptor, string eventArgs)
+        public async ValueTask DispatchBrowserEvent(string eventDescriptor, string eventArgs)
         {
-            if (CircuitHost == null)
+            var circuitHost = await GetActiveCircuitAsync();
+            if (circuitHost == null)
             {
-                Log.CircuitHostNotInitialized(_logger);
-                _ = NotifyClientError(Clients.Caller, "Circuit not initialized.");
                 return;
             }
 
             _ = CircuitHost.DispatchEvent(eventDescriptor, eventArgs);
         }
 
-        /// <summary>
-        /// Intended for framework use only. Applications should not call this method directly.
-        /// </summary>
-        public void OnRenderCompleted(long renderId, string errorMessageOrNull)
+        public async ValueTask OnRenderCompleted(long renderId, string errorMessageOrNull)
         {
-            if (CircuitHost == null)
+            var circuitHost = await GetActiveCircuitAsync();
+            if (circuitHost == null)
             {
-                Log.CircuitHostNotInitialized(_logger);
-                NotifyClientError(Clients.Caller, "Circuit not initialized.");
                 return;
             }
 
             Log.ReceivedConfirmationForBatch(_logger, renderId);
             CircuitHost.Renderer.OnRenderCompleted(renderId, errorMessageOrNull);
         }
 
-        public void OnLocationChanged(string uri, bool intercepted)
+        public async ValueTask OnLocationChanged(string uri, bool intercepted)
         {
-            if (CircuitHost == null)
+            var circuitHost = await GetActiveCircuitAsync();
+            if (circuitHost == null)
             {
-                Log.CircuitHostNotInitialized(_logger);
-                NotifyClientError(Clients.Caller, "Circuit not initialized.");
                 return;
             }
 
             _ = CircuitHost.OnLocationChangedAsync(uri, intercepted);
         }
 
+        private async ValueTask<CircuitHost> GetActiveCircuitAsync([CallerMemberName] string callSite = "")
+        {
+            var circuitHost = CircuitHost;
+            if (circuitHost == null)
+            {
+                // This can occur when a circuit host does not exist anymore due to an unhandled exception.
+                // We can reject this and terminate the connection.
+                Log.CircuitHostNotInitialized(_logger, callSite);
+                await NotifyClientError(Clients.Caller, "Circuit not initialized.");
```

Comment:
First 👍 on the whole `GetActiveCircuitAsync` pattern. That's really nice and clear.

Second, `"Circuit not initialized."` is a strange message if the scenario is "host does not exist anymore due to an unhandled exception". It would be good if we specified the difference between "circuit has never been initialized" and "circuit was already aborted", especially for the server-side log entry that developers will use for diagnosis in production.

Reply:
Good point. We should be able to do this with `CircuitHandle`.